scipy/stats/_stats_py.py

```diff
@@ -2821,26 +2821,35 @@ def obrientransform(*samples):
     that the variances are different.
 
     """
-    TINY = np.sqrt(np.finfo(float).eps)
+    xp = array_namespace(*samples)
 
     # `arrays` will hold the transformed arguments.
     arrays = []
     sLast = None
+    dtype = None
 
     for sample in samples:
-        a = np.asarray(sample)
-        n = len(a)
-        mu = np.mean(a)
+        a = xp.asarray(sample)
+        if xp.isdtype(a.dtype, 'integral'):
+            default_float = xp.asarray(1.).dtype
+            a = xp.asarray(a, dtype=default_float)
+        n = xp_size(a)
+        mu = xp.mean(a)
         sq = (a - mu)**2
-        sumsq = sq.sum()
+        sumsq = xp.sum(sq)
 
         # The O'Brien transform.
         t = ((n - 1.5) * n * sq - 0.5 * sumsq) / ((n - 1) * (n - 2))
 
         # Check that the mean of the transformed data is equal to the
         # original variance.
         var = sumsq / (n - 1)
-        if abs(var - np.mean(t)) > TINY:
+        difference = var - xp.mean(t)
+        # avoid recomputing `TINY` if not required
```

Comment:
The logic OK, but I'm not sure if it's saving us anything but a square root, at least with NumPy?

> The initial calculation of these parameters is expensive and negatively impacts import times. These objects are cached, so calling finfo() repeatedly inside your functions is not a problem.

Given the requirement to check `xp.isdtype`, it might not be worth the complexity. Did you compare the performance?